### ISSUE TITLE
S3 state bucket default encryption using SSE-KMS, but objects still encrypted with SSE-S3?

### ISSUE THREAD

**Issue by jpalomaki** [gruntwork-io/terragrunt#2293]
Hello,

I tested terragrunt `0.38.12` (on terraform `0.12.31`) in one of our AWS environments, applied the new state bucket settings as terragrunt prompted me to do so (`WARN[0003]   - Bucket Server-Side Encryption`, per the new feature to enforce those bucket settings, that appeared in `v0.37.0`), so that the bucket was in fact updated to use default encryption of `SSE-KMS` with the default `aws/s3` key.

Now when I apply a new config (`terragrunt apply`), the state file is encrypted with `SSE-S3`, instead of `SS3-KMS`. Is this expected? I presume this would happen if the state saving S3 PUT request specifies `SSE-S3` instead (overriding the bucket default of `SSE-KMS`).

I am currently not specifying a KMS key id in the remote state settings, so looking at https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#remote_state, I expected that to default to `SS3-KMS` (with key `aws/s3`), but it looks like we are getting `SSE-S3` instead.

Any pointers are greatly appreciated.

Thanks,
Jukka

**Comment by jpalomaki** [comment]
Furthermore, if we specify:

```hcl
remote_state {
  backend = "s3"

  config = {
    encrypt              = true
    ...
    bucket_sse_algorithm = "AES256"
  }
}
```
and we apply with that config, we still get the `WARN` to update the bucket SSE settings on every plan.

Looking at https://github.com/gruntwork-io/terragrunt/blob/3a23346e08ee574a4a3a23112350ea625d577a02/remote/remote_state_s3.go#L1080 terragrunt seems to only accept `aws:kms` for the algorithm, to detect SSE being enabled?

Is the use of `SSE-S3` discouraged here?

I myself find the terragrunt default bucket SSE settings vs what is actually being used by terraform somewhat confusing.

**Comment by jpalomaki** [comment]
> Hi, state file, in the end, is managed by `terraform` which encrypts objects with different encryption

All right, that explains the behavior.

**Comment by kevcube** [comment]
@denis256 I believe this is a bug. I am using the latest versions of terragrunt, terraform, AWS provider.

I create a terraform state bucket outside of terragrunt with SSE-S3 encryption enabled, I store the state for this bucket in itself, I then plan in terragrunt and terragrunt tells me I still need to apply encryption configuration, and if I say `y` then it will apply KMS encryption configuration instead of SSE-S3/AES256. Am I using `bucket_sse_algorithm` incorrectly? Or is it not being respected in terragrunt for some reason?

`root.hcl`

```
remote_state {
  backend = "s3"
  generate = {
    path      = "remote_state.tf"
    if_exists = "overwrite_terragrunt"
  }
  config = {
    region               = "us-west-2"
    bucket               = "defensestorm-terraform-new"
    key                  = "${path_relative_to_include()}/terraform.tfstate"
    encrypt              = true
    bucket_sse_algorithm = "AES256"
  }
}
```

`terragrunt.hcl`
```

# defaults for this cloudposse module are very good already, including server side encryption of the bucket
inputs = {
  allow_encrypted_uploads_only = true
  allow_ssl_requests_only      = true

  bucket_name        = "terraform-new"
}

terraform {
  source = "git@github.com:cloudposse/terraform-aws-s3-bucket.git?ref=3.0.0"
}

include "root" {
  path           = find_in_parent_folders("root.hcl")
  expose         = true
  merge_strategy = "deep"
}
```

```
kevcube@computer s3 % terragrunt plan         
WARN[0000] No double-slash (//) found in source URL /cloudposse/terraform-aws-s3-bucket.git. Relative paths in downloaded Terraform code may not work. 
WARN[0005] The remote state S3 bucket terraform-new needs to be updated:  prefix=[/Users/kevcube/terragrunt/s3] 
WARN[0005]   - Bucket Server-Side Encryption             prefix=[/Users/kevcube/terragrunt/s3] 
WARN[0005]   - Bucket Root Access                        prefix=[/Users/kevcube/terragrunt/s3] 
WARN[0005]   - Bucket Enforced TLS                       prefix=[/Users/kevcube/terragrunt/s3]
Remote state S3 bucket terraform-new is out of date. Would you like Terragrunt to update it? (y/n) n
```

**Comment by denis256** [comment]
Hi,
I will take a look on implementation and some more tests

**Comment by kevcube** [comment]
@denis see #2337 please

I was wrong, the bucket doesn't actually change the encryption from `AES256` to `aws:kms` after we've set the `bucket_sse_algorithm`, however it does still prompt us to update it on every run.

It just updates it to the same thing that it is already set to.

**Comment by kevcube** [comment]
@bradonkanyid yeah that's what I have addressed in #2337, hoping for a review soon. Otherwise you can just say "yes" to bucket changes and nothing will happen, but this is confusing my team.

cc @denis256

**Comment by sylwit** [comment]
Fix has been merged in version 0.43.2. It solved the issue. Thanks @kevcube

**Comment by number21** [comment]
Hi all!
I've tried terragrunt version v0.44.5 with enabled properties `bucket_sse_algorithm = "aws:kms"` and `bucket_sse_kms_key_id = my_kms_key` and encryption on state file present as SSE-S3, not SSE-KMS.
Default encryption for the bucket is SSE-KMS and my_kms_key.
Even if I manually change the encryption of tf-state file to KMS and the proper key, after applying tf-state encryption is rollback to SSE-S3.
I'm trying to change encryption for tf-state files from SSE-S3 to SSE-KMS and customer-managed key.
Is the above relevant for this tread?

**Comment by jpalomaki** [comment]
> Hi all! I've tried terragrunt version v0.44.5 with enabled properties `bucket_sse_algorithm = "aws:kms"` and `bucket_sse_kms_key_id = my_kms_key` and encryption on state file present as SSE-S3, not SSE-KMS. Default encryption for the bucket is SSE-KMS and my_kms_key. Even if I manually change the encryption of tf-state file to KMS and the proper key, after applying tf-state encryption is rollback to SSE-S3. I'm trying to change encryption for tf-state files from SSE-S3 to SSE-KMS and customer-managed key. Is the above relevant for this tread?

See https://github.com/gruntwork-io/terragrunt/issues/2293#issuecomment-1263985974

**Comment by jpalomaki** [comment]
I believe the main concerns and problems mentioned in this issue have been addressed, so I am closing this issue. Thanks everybody
